File: packages/java-client/wrapper/src/main/java/com/consentshield/sdk/ConsentVerifyException.java
package com.consentshield.sdk;

/**
 * Compliance-critical wrap thrown by verify calls when the SDK is fail-CLOSED
 * (the default) and the upstream cannot be reached after the retry budget is
 * exhausted (5xx, transport failure, or per-attempt timeout).
 *
 * <p>Fail-CLOSED is the safe default for verify: when in doubt, the caller
 * MUST gate the user out (HTTP 503 in a marketing flow, for example) rather
 * than treat the failure as implicit consent. To opt into fail-OPEN, set
 * {@code consentshield.fail-open=true} (Spring) or
 * {@code CONSENT_VERIFY_FAIL_OPEN=true} (env). Fail-OPEN does NOT throw —
 * it returns an {@link ConsentVerifyOutcome} carrying a {@code cause}
 * discriminator so the caller can record the override.
 *
 * <p>4xx responses are never wrapped here; they always surface as
 * {@link ConsentShieldApiException}.
 */
public final class ConsentVerifyException extends ConsentShieldException {
    private final FailureCause cause;
    private final String traceId;

    public ConsentVerifyException(FailureCause cause, String message, Throwable rootCause, String traceId) {
        super(message, rootCause);
        this.cause = cause;
        this.traceId = traceId;
    }

    public FailureCause getCause2() {
        return cause;
    }

    @Override
    public String getTraceId() {
        return traceId;
    }

    public enum FailureCause {
        /** Upstream returned 5xx after retries were exhausted. */
        SERVER_ERROR,
        /** Per-attempt timeout — never retried. */
        TIMEOUT,
        /** Transport failure (connection refused, DNS, TLS). */
        NETWORK,
    }
}

File: packages/php-client/wrapper/src/Exception/ConsentVerifyException.php
<?php declare(strict_types=1);

namespace ConsentShield\Sdk\Exception;

/**
 * Compliance-critical wrap thrown by verify calls when fail-CLOSED (default)
 * and the upstream cannot be reached after the retry budget is exhausted
 * (5xx, transport failure, or per-attempt timeout). 4xx responses are NEVER
 * wrapped here; they always surface as ConsentShieldApiException.
 */
final class ConsentVerifyException extends ConsentShieldException
{
    public function __construct(
        private readonly VerifyFailureCause $cause,
        string $message,
        ?string $traceId = null,
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, $traceId, $previous);
    }

    public function getCauseCode(): VerifyFailureCause
    {
        return $this->cause;
    }
}
